### DISCUSSION TITLE
Cannot use "authorization: Bearer" when making fileOperations.redirect requests with Minio

### DISCUSSION THREAD

**Discussion by vgwidt** [outline/outline#4005] [Self Hosting] [answered]
This is specifically "Bearer" with a capital B.  Lowercase "bearer" works fine.  This is because Minio will reject requests with multiple authentication requests (), and using "Bearer" will trip up [this check](https://github.com/minio/minio/blob/2070c215a298405b8606be1a8215e26d47a9653a/cmd/generic-handlers.go#L371).  

The check for AWS Signature V4 adds 1, and the user of "Bearer" will then trip up [this check](https://github.com/minio/minio/blob/2070c215a298405b8606be1a8215e26d47a9653a/cmd/auth-handler.go#L47), (const jwtAlgorithm is set to "Bearer" [here](https://github.com/minio/minio/blob/0ea5c9d8e88197dc87dfe36184dd808569929746/cmd/jwt.go#L36)), adding another +1 resulting in the error 

I used both Insomnia and the Rust reqwest crate to make POST requests (fileOperations.redirect) to test the issue.  I verified using mc admin trace that the requests are virtually identical except for the capitalization of "B".  Here is the failure message if you use the capital "B":
```
<Error><Code>InvalidRequest</Code><Message>Invalid Request (request has multiple authentication types, please use one)</Message> ...
```

Reproduce by using Minio, making a POST request to /api/fileOperations.redirect with the id of the file in the body, using the following headers:
```
Content-Type: application/json
accept: application/json
authorization: Bearer insert_key
```

I tested it using AWS S3 and could not reproduce, it works fine either way.

Posting this as a discussion to start, as I don't know which side the issue is on.  It isn't the biggest deal as it looks to only be Minio and can be worked around using a lower-case "b", but the documentation also suggests using an upper-case "B" (https://www.getoutline.com/developers).

**Comment by tommoor** [comment]
I'd suggest you file this as a bug on Minio, the header should be case insensitive and the fact that the behavior is different depending on casing puts it firmly in bug territory.

**Reply by vgwidt** [reply]
Thanks @tommoor, I'll do that.

**Reply by vgwidt** [reply]
@tommoor question for you, but should Minio be able to see those headers or are they important?  It says it is a GET request and I wonder if Minio should be able to see the Outline API key.  Here is an example of what Minio sees:
```
{
 "type": "S3",
 "host": "xxx.xxx.xxx.xxx",
 "api": "s3.GetObject",
 "time": "2022-08-24T07:11:10.723802596Z",
 "duration": 173600619,
 "path": "/bucket/xxx/xxx/xxx.zip",
 "error": "",
 "request": {
  "time": "2022-08-24T07:11:10.723802596Z",
  "proto": "HTTP/1.1",
  "method": "GET",
  "path": "/bucket/xxx/xxx/xxx.zip",
  "rawQuery": "X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=xxx&X-Amz-Date=20220824T071110Z&X-Amz-Expires=60&X-Amz-Signature=xxx&X-Amz-SignedHeaders=host&response-content-disposition=attachment",
  "headers": {
   "Accept": "application/json",
   "Authorization": "bearer xxx",
   "Connection": "upgrade",
   "Content-Length": "0",
   "Content-Type": "application/json",
   "Host": "xxx.xxx.xxx.xxx",
   "User-Agent": "insomnia/2022.5.1",
   "X-Forwarded-For": "xxx.xxx.xxx.xxx",
   "X-Forwarded-Proto": "http",
   "X-Real-Ip": "xxx.xxx.xxx.xxx
  },
  "body": "<BODY>"
 },
``

**Reply by tommoor** [reply]
Weird, it feels like insomnia might be doing something here but it is concerning that authorization would ever be sent to Minio.

The Outline API responds with a standard 302 redirect – it's not even _possible_ to include any headers that were sent to the API such as authorization.

**Reply by vgwidt** [reply]
I had the same result using reqwest too, can’t remember what I saw for the actual requests from Outline though.  I’ll be back in front of a computer in 12 hours to double check.

**Reply by vgwidt** [reply]
It does look to be decided by the client where the headers are kept on a redirect ([Chrome Example](https://stackoverflow.com/questions/35400943/how-to-remove-authorization-header-in-a-http-302-response), [Axios Example](https://github.com/axios/axios/issues/2855)).  It looks like these headers are generally kept if the redirect is [on the same domain](https://stackoverflow.com/questions/72002362/react-cors-error-when-api-302-redirects-to-same-domain) whereas the client should not forward them if it is a separate domain.  This explains why it fails in Minio and not AWS in my case despite AWS appearing to have the same check.  I'll verify by sticking Minio on a different domain but I think this behavior is expected and may be a non-issue.  Sorry for the trouble :sweat_smile:

**Reply by vgwidt** [reply]
Confirmed the headers are not passed as long as S3 storage is not on the same domain.

**Comment by vgwidt** [comment, marked as answer]
The issue is that headers specific to Outline should not be passed to Minio.  However, some clients will pass the headers if both Outline and Minio are on the same domain.  Putting Minio on a separate domain will result in the headers being dropped making this a non-issue.

**Reply by tommoor** [reply]
Great, I've added this to the troubleshooting doc for others
